### PR TITLE
[ENH] : Collapsed Sidebar Feature for all Screen 

### LINKED ISSUE #25: Feature Request: Collapsible Slide Bar for Frontend Chatbot Interface
A feature to add a collapsible slide bar in the frontend chatbot interface. This will enhance user experience by allowing users to expand or collapse the side bar as needed, providing a cleaner and more focused interface.

## Requirements
- A slide bar on the left side of the chatbot interface.
- The slide bar should be collapsible with a toggle button.
- When collapsed, the slide bar should hide completely or minimize to a small icon.
- Smooth animation for expanding and collapsing the slide bar.
- Ensure the feature is responsive and works well on both desktop and mobile devices.

## Acceptance Criteria
- [x] A collapsible slide bar is present in the frontend chatbot interface.
- [x] The slide bar can be toggled open and closed by a button.
- [x] Smooth animations for the slide bar's collapsing and expanding actions.
- [x] The slide bar does not interfere with the chatbot's main functionality.
- [x] The feature is fully responsive on various screen sizes.
- [x] The UI/UX design follows the project's style guide.

## Mockups / Examples
- [x] Include any relevant mockups or examples here to visualize the feature.

### PR DESCRIPTION
**Pull request by Sauradip07**
- Closes #25 

Changes proposed in this pull request:

- The sidebar is collapsed by default in this pull request for all screen sizes (big, medium, and tiny).
- For a medium and small screen, it is collapsed; for a large screen, it is open. 

## Checklist
_This section is for the PR reviewer_

- [x] PR has an interpretable title with a prefix (`[ENH]`, `[FIX]`, `[REF]`, `[TST]`, `[CI]`, `[MNT]`, `[INF]`, `[MODEL]`, `[DOC]`) (see our [Contributing Guidelines](/docs/pull_request_guidelines.md) for more info)

- [x] PR links to GitHub issue with mention `Closes #XXXX`

https://github.com/user-attachments/assets/46846a0a-b654-4dd7-95db-6d89ac83ded9

**Please advise if any modifications are required.**

### CODE REVIEW

File: src/frontend/components/sidebar.tsx [hunk 1]
```diff
@@ -1,14 +1,55 @@
...
-    return (
-        <aside className="sticky top-0 overflow-y-auto h-full bg-primary text-white py-1 px-2 w-64 md:block hidden">
-            <MenuOptions />
+  const [sidebarOpen, setSideBarOpen] = useState(true);
+
+  const handleViewSidebar = () => {
+    setSideBarOpen(!sidebarOpen);
+  };
+
+  useEffect(() => {
```
Comment (swaptr) [inline src/frontend/components/sidebar.tsx:18]: You can utilize the `hooks/useWindowSize` hook to clean the code here.

File: src/frontend/components/sidebar.tsx [hunk 2]
```diff
@@ -1,14 +1,55 @@
...
+      >
+        <MenuOptions isOpen={sidebarOpen} toggleSidebar={handleViewSidebar} />
+      </aside>
+      {!sidebarOpen && (
+        <aside className="h-full w-full flex items-center justify-center bg-transparent">
+          <button
+            onClick={handleViewSidebar}
+            className="px-1 py-2 bg-blue-500 text-white rounded hover:bg-blue-600 transition-all duration-300"
+          >
+            <IoIosArrowForward size={25} />
```
Comment (swaptr) [inline src/frontend/components/sidebar.tsx:47]: You can use something from the `lucide-react` icons gallery here: https://lucide.dev/icons/. Later on we will have our own limited icon set and can remove all icons related dependencies.

File: src/frontend/components/sidebar.tsx [hunk 3]
```diff
@@ -1,14 +1,55 @@
...
+    <div>
+      <aside
+        className={`fixed top-0 h-full bg-primary text-white py-1 px-2 w-64 transition-transform duration-300 ease-in-out ${
+          sidebarOpen ? 'translate-x-0' : '-translate-x-full'
+        }`}
+      >
+        <MenuOptions isOpen={sidebarOpen} toggleSidebar={handleViewSidebar} />
+      </aside>
+      {!sidebarOpen && (
+        <aside className="h-full w-full flex items-center justify-center bg-transparent">
```
Comment (swaptr) [inline src/frontend/components/sidebar.tsx:42]: This component occupies viewport space and shifts the entire <main>, which is particularly problematic on smaller screens. Consider moving it as an overlay or placing the button in the top bar next to the product title.

**Comment by swaptr** [comment]
Thank you for submitting the PR, it looks very promising. However, this PR is failing the DCO check. Please sign your commits to resolve this issue.

File: src/frontend/components/sidebar.tsx (hunk 1 above)
Comment (Sauradip07) [reply src/frontend/components/sidebar.tsx:20]: Done

File: src/frontend/components/sidebar.tsx (hunk 2 above)
Comment (Sauradip07) [reply src/frontend/components/sidebar.tsx]: done

File: src/frontend/components/sidebar.tsx (hunk 3 above)
Comment (Sauradip07) [reply src/frontend/components/sidebar.tsx]: Done 
Screenshot from 2024-07-19 21-06-57

**Comment by swaptr** [comment]
I have gone through this PR and I found couple of bugs:
1. On portrait mode, the main content is no longer accessible.
2. There is a very slight visual bug with the collapse/expand buttons, if you look closely the collapse button is visible as soon as the sidebar sheet is visible and it overlays on top of it. (see video)

bug

3. [REQUEST] Would it be possible to have a collapsible sidebar in landscape mode and on mobile it transforms into a sheet?

**Comment by Sauradip07** [comment]
> I have gone through this PR and I found couple of bugs:
> 
> 1. On portrait mode, the main content is no longer accessible. (EDIT: The issue is not on portrait mode but when the sheet is closed).
> 2. There is a very slight visual bug with the collapse/expand buttons, if you look closely the collapse button is visible as soon as the sidebar sheet is visible and it overlays on top of it. (see video)
> 
> bug
> 
> 3. [REQUEST] Would it be possible to have a collapsible sidebar in landscape mode and on mobile it transforms into a sheet?

**Fix it, just by adding the delay into it**
new

**Comment by swaptr** [comment]
Bug: On mobile/portrait mode when I try to load the application there is a Hydration error on the browser.
How to reproduce this bug: Keep the browser in portrait mode, or zoom till it triggers portrait mode and reload/refresh the page.
Tested on: Firefox 128.0, Chromium 126.0
Log:
```
Unhandled Runtime Error

Error: Hydration failed because the initial UI does not match what was rendered on the server.
See more info here: https://nextjs.org/docs/messages/react-hydration-error

Expected server HTML to contain a matching <div> in <div>.

Call Stack (React):
throwOnHydrationMismatch
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (6981:1)
tryToClaimNextHydratableInstance
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (7040:1)
updateHostComponent$1
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (16621:1)
beginWork$1
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (18503:1)
callCallback
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (20565:1)
invokeGuardedCallbackImpl
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (20614:1)
invokeGuardedCallback
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (20689:1)
beginWork
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (26949:1)
performUnitOfWork
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (25748:1)
workLoopConcurrent
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (25734:1)
renderRootConcurrent
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (25690:1)
performConcurrentWorkOnRoot
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (24504:1)
workLoop
node_modules/next/dist/compiled/scheduler/cjs/scheduler.development.js (256:1)
flushWork
node_modules/next/dist/compiled/scheduler/cjs/scheduler.development.js (225:1)
performWorkUntilDeadline
node_modules/next/dist/compiled/scheduler/cjs/scheduler.development.js (534:1)
```

File: src/frontend/components/sidebar.tsx [hunk 4]
```diff
@@ -1,14 +1,107 @@
...
+  SheetHeader,
+  SheetTitle,
+  SheetTrigger,
+} from "@/components/ui/sheet";
 
-type Props = {}
+type Props = {
+  isOpen: boolean;
+  toggleSidebar: () => void;
+};
```
Comment (swaptr) [inline src/frontend/components/sidebar.tsx:20]: I see you added these props but it doesnt seem they are used anywhere. Also when we call `Sidebar` on `app/page.tsx`, it doesn't seem like we pass these props.

File: src/frontend/components/sidebar.tsx [hunk 5]
```diff
@@ -1,14 +1,114 @@
...
+  const [isMounted, setIsMounted] = useState(false);
+  const { isMobile } = useWindowSize({ initializeWithValue: true });
+
+  const handleViewSidebar = () => {
+    setSideBarOpen(!sidebarOpen);
+  };
+
+  useEffect(() => {
+    setIsMounted(true);
+  },[]);
```
Comment (swaptr) [inline src/frontend/components/sidebar.tsx:35]:
```suggestion
  },[isMounted]);
```

**Review by swaptr — CHANGES_REQUESTED** [review]
LGTM. Just this fix and it should be good to land. Thanks!

**Pull request merged**